libcoz/libcoz.cpp

```diff
@@ -225,6 +227,26 @@ static bool is_coz_signal(int signum) {
   return signum == SampleSignal || signum == SIGSEGV || signum == SIGABRT;
 }
 
+namespace {
+
+template <size_t Size>
+class LocalArena {
+  std::array<char, Size> _data;
+  std::atomic<char *> _offset{_data.begin()};
+
+public:
+  void * allocate(size_t bytes_requested) {
+    bytes_requested = (bytes_requested + 8) & ((~0ull) << 3);
```

Comment:
Add 7, not 8. You're wasting 8 bytes when address is already aligned.

Also, 8, which is `sizeof(double)` which on most platform is the most demanding type when it comes to alignment, might not always be enough, as far as I know. For example, for some type that are used to SIMD instructions.
If I'm not mistaken, you should use `sizeof(max_align_t)`, which is 32 on my computer.

Reply:
Alignment is a point of controversy, but I think you meant `alignof(max_align_t)`. 32 bytes is definitely overkill!

Reply:
Yes, I meant `alignof(max_align_t)`. I've actually made the same mistake in my PR. Last time I implemented my own malloc, I had to use C++03, although C++14 was already available, so I'm probably a bit stuck in that time when it comes to `alignof`. :)
Anyway, they both return the same value in case of probably all scalar types, and they certainly do in this case. But I should fix it, because of correct semantics.
I don't know if it is point of controversy or not, this [doc](https://en.cppreference.com/w/cpp/types/max_align_t) suggests that malloc should return addresses aligned to `alignof(std::max_align_t)`, and I kind of don't want to get into details of different standards right now. :)
On my machine, which is Ubuntu 18.04 running on high end Intel processor, this value is 32, and  glibc malloc really returns addresses aligned to 32 every time.

This is how it is defined:
```
typedef struct {
  long long __max_align_ll __attribute__((__aligned__(__alignof__(long long))));
  long double __max_align_ld __attribute__((__aligned__(__alignof__(long double))));
  /* _Float128 is defined as a basic type, so max_align_t must be
     sufficiently aligned for it.  This code must work in C++, so we
     use __float128 here; that is only available on some
     architectures, but only on i386 is extra alignment needed for
     __float128.  */
#ifdef __i386__
  __float128 __max_align_f128 __attribute__((__aligned__(__alignof(__float128))));
#endif
} max_align_t;
```

So I guess, it might be OK not to align less than system's malloc aligns.

Reply:
Valid point, thanks!